### ISSUE TITLE
False positive: sanitize_html_class() incorrectly flagged for missing unslash

### ISSUE THREAD

**Issue by IanDelMar** [WordPress/WordPress-Coding-Standards#2516]
## Bug Description
When using `sanitize_html_class()` to sanitize an HTML class, PHP_CodeSniffer reports that the passed variable is not unslashed before sanitization. However, `sanitize_html_class()` strips the string down to `A-Z`, `a-z`, `0-9`, `_`, and `-` ([code reference](https://developer.wordpress.org/reference/functions/sanitize_html_class/)), and therefore it should be listed as an [unslashing sanitizing function](https://github.com/WordPress/WordPress-Coding-Standards/blob/c028f329aadd97788288c5cb4d1884eca50b954d/WordPress/Helpers/SanitizationHelperTrait.php#L139-L149), similar to `sanitize_key()`. Instead, it is currently listed as a general [sanitizing function](https://github.com/WordPress/WordPress-Coding-Standards/blob/c028f329aadd97788288c5cb4d1884eca50b954d/WordPress/Helpers/SanitizationHelperTrait.php#L94).

## Minimal Code Snippet

The issue happens when running this command:
```bash
phpcs --standard=WordPress --sniffs=WordPress.Security.ValidatedSanitizedInput test.php
```

... over a file containing this code:
```php
if (! isset($_GET['html_class']) || ! is_string($_GET['html_class'])) {
	return;
}

$htmlClass = sanitize_html_class($_GET['html_class']);
```

## Error Code
WordPress.Security.ValidatedSanitizedInput.MissingUnslash

## Environment

| Question                 | Answer
| ------------------------ | -------
| PHP version              | 8.2.4
| PHP_CodeSniffer version  | 3.11.2
| WordPressCS version      | 3.1.0
| PHPCSUtils version       | 1.0.12
| PHPCSExtra version       | 1.2.1
| WordPressCS install type | Composer project local
| IDE (if relevant) | N/A

## Tested Against `develop` Branch?
- [x] I have verified the issue still exists in the `develop` branch of WordPressCS.

**Comment by rodrigoprimo** [comment]
Thanks for reporting this issue, @IanDelMar. I agree with you that `sanitize_html_class()` should be moved to `$unslashingSanitizingFunctions`. Would you like to open a PR?

**Comment by rodrigoprimo** [comment]
@IanDelMar, since I haven't heard back from you, I'd like to go ahead and open a PR for this issue. Let me know if you'd still prefer to work on it yourself, and I'm happy to wait.

Before opening the PR, I'd like to get input from the other maintainers on the scope of the change. I noticed other functions currently on `$sanitizingFunctions` also implicitly strip backslashes as far as I can see, so they could be moved to `$unslashingSanitizingFunctions` as well. Below are the ones that I believe should be moved to `$unslashingSanitizingFunctions`:

- [`esc_url_raw()`](https://github.com/WordPress/wordpress-develop/blob/51a5f4d5fdcc5a6b1e859f05f309b3339f455e0a/src/wp-includes/formatting.php#L4656) / [`sanitize_url()`](https://github.com/WordPress/wordpress-develop/blob/51a5f4d5fdcc5a6b1e859f05f309b3339f455e0a/src/wp-includes/formatting.php#L4676): delegates to [`esc_url()`](https://github.com/WordPress/wordpress-develop/blob/51a5f4d5fdcc5a6b1e859f05f309b3339f455e0a/src/wp-includes/formatting.php#L4538), whose regex `[^a-z0-9-~+_.?#=!&;,/:%@$|*'()\[\]\x80-\xff]` strips backslashes.
- [`sanitize_email()`](https://github.com/WordPress/wordpress-develop/blob/51a5f4d5fdcc5a6b1e859f05f309b3339f455e0a/src/wp-includes/formatting.php#L3818): the regex ``[^a-zA-Z0-9!#$%&'*+/=?^_`{|}~.-]`` strips backslashes.
- [`sanitize_file_name()`](https://github.com/WordPress/wordpress-develop/blob/51a5f4d5fdcc5a6b1e859f05f309b3339f455e0a/src/wp-includes/formatting.php#L2035): `\` is explicitly included in the `$special_chars` array passed to `str_replace()`.
- [`sanitize_hex_color()`](https://github.com/WordPress/wordpress-develop/blob/51a5f4d5fdcc5a6b1e859f05f309b3339f455e0a/src/wp-includes/formatting.php#L6306): returns the input only if it matches `^#([A-Fa-f0-9]{3}){1,2}$`, otherwise returns nothing.
- [`sanitize_hex_color_no_hash()`](https://github.com/WordPress/wordpress-develop/blob/51a5f4d5fdcc5a6b1e859f05f309b3339f455e0a/src/wp-includes/formatting.php#L6333): delegates to `sanitize_hex_color()`.
- [`sanitize_html_class()`](https://github.com/WordPress/wordpress-develop/blob/51a5f4d5fdcc5a6b1e859f05f309b3339f455e0a/src/wp-includes/formatting.php#L2437): regex `[^A-Za-z0-9_-]` strips backslashes.
- [`sanitize_mime_type()`](https://github.com/WordPress/wordpress-develop/blob/51a5f4d5fdcc5a6b1e859f05f309b3339f455e0a/src/wp-includes/formatting.php#L5783): regex `[^-+*.a-zA-Z0-9\/]` strips backslashes.
- [`sanitize_sql_orderby()`](https://github.com/WordPress/wordpress-develop/blob/51a5f4d5fdcc5a6b1e859f05f309b3339f455e0a/src/wp-includes/formatting.php#L2415): returns the input only if it matches one of two regexes (neither of which allows backslashes), `false` otherwise.
- [`sanitize_title_with_dashes()`](https://github.com/WordPress/wordpress-develop/blob/51a5f4d5fdcc5a6b1e859f05f309b3339f455e0a/src/wp-includes/formatting.php#L2281): final regex `[^%a-z0-9 _-]` strips backslashes.
- [`wp_sanitize_redirect()`](https://github.com/WordPress/wordpress-develop/blob/51a5f4d5fdcc5a6b1e859f05f309b3339f455e0a/src/wp-includes/pluggable.php#L1553): regex `[^a-z0-9-~+_.?#=&;,/:%!*\[\]()@]` strips backslashes.
- [`sanitize_title()`](https://github.com/WordPress/wordpress-develop/blob/51a5f4d5fdcc5a6b1e859f05f309b3339f455e0a/src/wp-includes/formatting.php#L2227) — the function body itself doesn't strip backslashes, but it calls `apply_filters( 'sanitize_title', ... )` and `sanitize_title_with_dashes()` is hooked to that filter by default.
- [`sanitize_title_for_query()`](https://github.com/WordPress/wordpress-develop/blob/51a5f4d5fdcc5a6b1e859f05f309b3339f455e0a/src/wp-includes/formatting.php#L2262): wraps `sanitize_title( $title, '', 'query' )`, so same reasoning as `sanitize_title()`.

Commenting here to get input from others in case I'm missing something regarding the functions above before proceeding.

**Comment by IanDelMar** [comment]
Sorry for not getting back earlier. Yes, please go ahead and open a PR for this issue.
